# Homework 03 - Python-Daniel Teeuwsen

In [1]:
import numpy as np
from scipy.special import factorial

#### Excercise 1
A ball is dropped from a tower of height $h$ with initial velocity zero.  Write a program that asks the user to enter the height in meters of the tower and then calculates and prints the time the ball takes until it hits the ground, ignoring air resistance.  Use your program to calculate the time for a ball dropped from a 100 m high tower. 

In [2]:
h = input('Please enter the height of the tower in meters: ')  # 0 = h-(gravity/2)*time^2 => t = sqrt(h*(2/g)) 
g = 9.80665 
t = np.round(np.sqrt(float(h)*(2/g)),decimals=2)
print('It takes approximately ' + str(t) + ' seconds for the ball to hit the ground from a height of ' + h +' meters.' )

Please enter the height of the tower in meters: 100
It takes approximately 4.52 seconds for the ball to hit the ground from a height of 100 meters.


The program calculated that it takes about 4.52 seconds for the ball to hit the ground from a height of 100 meters.

#### Excercise 2
A satellite is to be launched into a circular orbit around the Earth so that it orbits the planet once every $T$ seconds.
    
**A.** In a Markdown cell and with Latex equations, show that the altitude $h$ above the Earth's surface that the satellite must have is
$$
h = \biggl( {GMT^2\over4\pi^2} \biggr)^{1/3} - R,
$$ where $G=6.67\times10^{-11}\,\textrm{m}^3\,\textrm{kg}^{-1}\,\textrm{s}^{-2}$ is Newton's gravitational constant, $M=5.97\times10^{24}\textrm{ kg}\,$ is the mass of the Earth, and $R=6371\textrm{ km}\,$ is its radius.

**B.** In a Code cell, write a program that asks the user to enter the desired value of $T$ and then calculates and prints out the correct altitude in kilometers, to two decimal places.

**C.** Use your program to calculate the altitudes of satellites that orbit the Earth once a day (so-called "geosynchronous" orbit), once every 90  minutes, and once every 45 minutes.

**Answer**

### Part A:
By Newton's Law of Universal Gravitation we have that $ F=\frac{GmM}{r^2} $ where $m$ is the mass of the satellite and $r$ is the radius of the orbit. Then by the second law of motion $ F=ma $ and knowing that $ a=\frac{\hphantom{a}v^2}{r} $ we obtain $ F=m\frac{\hphantom{a}v^2}{r} $. Now, since we know that the satellite travels $ 2\pi\textit{r} $ once every $T$ seconds we can deduce that velocity $ v=\frac{2\pi\textit{r}}{T} $. Next, by equating the right side of both expressions and solving for $r$ we find that $ r=\left(\frac{GMT^2}{4\pi^2}\right)^{1/3} $ Finally, $r$ is the distance from the center of the Earth to the center of the satellite, and $R$ is the distance from the center of the Earth to the surface of the Earth. Therefore, the satellite's altitude $h$ above the Earth is $ r-R $ or in other words $ h= \left(\frac{GMT^2}{4\pi^2}\right)^{1/3} - R $.



In [3]:
T = input('Please enter the desired orbit time of the satellite in seconds: ')
G = 6.67*10**(-20) #converted meters^3 to kilometers^3 -11 to -20
M = 5.97*10**24
R = 6371
h = np.round( ((G*M*int(T)**2)/(4*np.pi**2))**(1/3) - R, decimals=2)
print('The altitude of the satellite must then be ' + str(h) + ' kilometers.')

Please enter the desired orbit time of the satellite in seconds: 86400
The altitude of the satellite must then be 35855.91 kilometers.


I)   A day has 24*60*60 = 86400 seconds, thus the altitude of a satellite orbiting once a day must then be 35855.91 kilometers.
II)  90*60 = 5400 seconds,thus the altitude of the satellite must then be 279.32 kilometers.
III) 45*60 = 2700 seconds, thus the altitude of the satellite must then be -2181.56 kilometers which is clearly impossible.

#### Excercise 3
The [Catalan numbers][catalan] $C_n$ are a sequence of integers 1, 1, 2, 5, 14, 42, 132... that play an important role in quantum mechanics and the theory of disordered systems.  (They were central to Eugene Wigner's proof of the so-called semicircle law.)  They are given by 
$$
C_0 = 1,\qquad C_{n+1} = {4n+2\over n+2}\,C_n.
$$
Write a program that prints in increasing order all Catalan numbers less than or equal to one billion.

[catalan]: https://en.wikipedia.org/wiki/Catalan_number

In [4]:
a = 1
n = 0
while a < 10**9:
    print(a)
    a = int((((4*n)+2)/(n+2))*a)
    n = n+1

1
1
2
5
14
42
132
429
1430
4862
16796
58786
208012
742900
2674440
9694845
35357670
129644790
477638700


#### Excercise 4

The [binomial coefficient][binomial] ${n\choose k}$ is an integer equal to
$$
{n\choose k} = {n!\over k!(n-k)!}
  = {n\times(n-1)\times(n-2)\times\ldots\times(n-k+1)\over
     1\times2\times\ldots\times k}
$$
when $k\ge1$, or ${n\choose0}=1$ when $k=0$.</font>


**A.** Using this form for the binomial coefficient, write a user-defined function `binomial(n,k)` that calculates the binomial coefficient for given $n$ and $k$.  Make sure your function returns the answer in the form of an integer (not a float) and gives the correct value of 1 for the case where $k=0$.

Use `scipy.special` to access the [`factorial`][factorial] function.

**B.** Using your function, write a program to print out the first 20 lines of [*Pascal's triangle*.][pascal]  The $n$-th line of Pascal's triangle contains $n+1$ numbers, which are the coefficients ${n\choose 0}$, ${n\choose1}$,   and so on up to ${n\choose n}$.  Thus the first few lines are

```python
1 1 
1 2 1 
1 3 3 1 
1 4 6 4 1
```

**C.** The probability that an unbiased coin, tossed $n$ times, will come up  heads $k$ times is ${n\choose k}/2^n$.  Write a program to calculate (a) the total probability that a coin tossed 100 times comes up heads exactly 60 times, and (b) the total probability that a coin tossed 100 times comes up heads 60 or more times.</font>

[binomial]: https://en.wikipedia.org/wiki/Binomial_coefficient
[pascal]: https://en.wikipedia.org/wiki/Pascal%27s_triangle
[factorial]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.factorial.html


<div class="alert alert-block alert-info">
    <b>Tip:</b> In the Pascal's Traingle, each line grows in the number of elements. If you are printing the elements one by one, and you <strong>do not</strong> want a <tt>print</tt> commmand to move to the next line autoamtically, use the option <tt>print(x,end='')</tt>.
    
When you want to insert a line break, just print a single space character; <tt>print(' ')</tt>
</div>

In [5]:
## Part A ##


def binomial(n,k):
    return int(factorial(n,exact=True)/(factorial(k,exact=True)*factorial(n-k,exact=True)))


# Some test cases #

print(binomial(5,1))
print(binomial(4,2))
print(binomial(3,0))
print(binomial(0,0))

## Test cases which should return errors ##

#print(binomial(3,4))
#print(binomial(3.4,2.2))
#print(binomial(-3,-4))
#print(binomial(-3,-2))

5
6
1
1


In [6]:
## Part B ##

k=20
for n in range(1,k+1):
    myString=''
    for i in range(n+1):
        myString = myString + ' ' + str(binomial(n,i))
    print(myString)


#The following code outputs pascal's triangle as an actual triangle (only works up to k=14 since it runs out of space)
print('\n')
k=14 
space = len(str(binomial(k,np.floor(k/2)))) #returns number of digits of largest value

for n in range(k):
    myString = (k-n)*space*' '
    for i in range(n+1):
        myString = myString + (space+space-len(str(binomial(n,i))))*' ' + str(binomial(n,i))
    print(myString)

 1 1
 1 2 1
 1 3 3 1
 1 4 6 4 1
 1 5 10 10 5 1
 1 6 15 20 15 6 1
 1 7 21 35 35 21 7 1
 1 8 28 56 70 56 28 8 1
 1 9 36 84 126 126 84 36 9 1
 1 10 45 120 210 252 210 120 45 10 1
 1 11 55 165 330 462 462 330 165 55 11 1
 1 12 66 220 495 792 924 792 495 220 66 12 1
 1 13 78 286 715 1287 1716 1716 1287 715 286 78 13 1
 1 14 91 364 1001 2002 3003 3432 3003 2002 1001 364 91 14 1
 1 15 105 455 1365 3003 5005 6435 6435 5005 3003 1365 455 105 15 1
 1 16 120 560 1820 4368 8008 11440 12870 11440 8008 4368 1820 560 120 16 1
 1 17 136 680 2380 6188 12376 19448 24310 24310 19448 12376 6188 2380 680 136 17 1
 1 18 153 816 3060 8568 18564 31824 43758 48620 43758 31824 18564 8568 3060 816 153 18 1
 1 19 171 969 3876 11628 27132 50388 75582 92378 92378 75582 50388 27132 11628 3876 969 171 19 1
 1 20 190 1140 4845 15504 38760 77520 125970 167960 184756 167960 125970 77520 38760 15504 4845 1140 190 20 1


                                                               1
                                     

In [7]:
## Part C ##

#i)

pOf60 = np.round(binomial(100,60)/(2**100),decimals=4)
print('The probability that 60 "heads" will be observed after tossing a fair coin 100 times is ' + str(pOf60) + ' or ' + str(pOf60*100) + '%.')


#ii)

mySum = 0
for i in range(60,101):
    mySum = mySum + binomial(100,i)/(2**100)
    
pOf60orMore = np.round(mySum,decimals=4)
print('The probability that the number of "heads" observed will be 60 or greater is ' + str(pOf60orMore) + ' or ' + str(np.round(pOf60orMore*100,decimals=2)) + '%.' )



#Alternative solution

#complement = 0
#for i in range(41):
#    complement = complement + (binomial(100,i)/(2**100))

#pOf60orMore = np.round(complement,decimals=4)
#print(pOf60orMore)

The probability that 60 "heads" will be observed after tossing a fair coin 100 times is 0.0108 or 1.08%.
The probability that the number of "heads" observed will be 60 or greater is 0.0284 or 2.84%.


In [ ]:
for i in range(20)